mattdrayer/SOL-465: Configure Certificate Web View via Studio

mattdrayer opened the pull request:
JIRA:  https://openedx.atlassian.net/browse/SOL-465

Sandboxes:
  * LMS 
    * http://pr7882.m.sandbox.edx.org  (log in as 'honor' user)
    * Click 'View Certificate' button on dashboard for edX Demonstration Course
  * Studio 
    * http://studio.pr7882.m.sandbox.edx.org (log in as 'staff' user)
    * Select Course > Settings > Certificates

cahrens commented:
@mattdrayer is this the PR related to https://openedx.atlassian.net/browse/SOL-704?

mattdrayer commented:
Yes, it's not quite ready but I wanted to get the PR up for staging purposes -- trying to wrap up the remaining test coverage and some small things, then will need to do our internal review.

cahrens commented:
Great. Our plan for this quarter is to only do a high-level (max 2 hour) review, so please do make sure a full review has been done first. If you have questions, check with Steve.

mattdrayer commented:
Jenkins test this please.

mattdrayer commented:
Jenkins, test this please.

mattdrayer commented on cms/templates/widgets/header.html:
```diff
@@ -98,6 +99,9 @@ <h3 class="title"><span class="label"><span class="label-prefix sr">${_("Course"
                   <li class="nav-item nav-course-settings-advanced">
                     <a href="${advanced_settings_url}">${_("Advanced Settings")}</a>
                   </li>
+                  <li class="nav-item nav-course-settings-certificates">
```
@asadiqbal08 -- we need to show this link only if the certificates web view has been enabled in configuration -- can you add a check for the "CERTIFICATES_HTML_VIEW" Django setting?

asadiqbal08 replied:
Sorry, I just saw this. 
Sure i will add the check also. I will update you in my upcoming PR.

mattdrayer commented:
Jenkins, test this please.

mattdrayer commented:
@martynjames @marcotuts -- here is corresponding PR for SOL-465 -- I'm going to move the ticket into the Internal Review column in JIRA now -- will create/assign subtasks for you, as well as @cptvitamin

mattdrayer commented:
@talbs @catong -- FYI on this PR

catong commented:
@mattdrayer Woooot and congrats!! Is there also a sandbox?

mattdrayer commented:
@cptvitamin -- FYI on this PR

mattdrayer commented:
@catong -- there is a sandbox available, yes -- http://pr7882.m.sandbox.edx.org -- it's up-to-date as of yesterday afternoon, so should have everything in terms of UI/UX at present.

catong commented:
Awesome; thanks!

catong commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+                    </span>
+                <% } %>
+                <input id="certificate-name-<%= uniqueId %>" class="collection-name-input input-text" name="certificate-name" type="text" placeholder="<%= gettext("This is the Name of the Certificate") %>" value="<%= name %>">
+                <span class="tip tip-stacked"><%= gettext("Name or short description of the configuration") %></span>
```
"the configuration" -> "the certificate"

cptvitamin replied:
can you add an ID here i.e. ``` id="<%= uniqueId %>-tip"``` and then add an ```aria-describedby="<%= uniqueId %>-tip"``` to the ```<input>``` above.  This programmatically links the help text to the control, so it will be read when the control has focus.

catong commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+                        <span class="certificate-value"><%= id %></span>
+                    </span>
+                <% } %>
+                <input id="certificate-name-<%= uniqueId %>" class="collection-name-input input-text" name="certificate-name" type="text" placeholder="<%= gettext("This is the Name of the Certificate") %>" value="<%= name %>">
```
I'm not seeing this placeholder text in the sandbox. I'm seeing "Default Name" instead.
The default text in fields on this page seems to be darker than what we use on other Studio pages, e.g. Group Configurations or Libraries.

catong commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+            </div>
+            <div class="input-wrap field text add-certificate-description">
+                <label for="certificate-description-<%= uniqueId %>"><%= gettext("Description") %></label>
+                <textarea id="certificate-description-<%= uniqueId %>" class="certificate-description-input text input-text" name="certificate-description" placeholder="<%= gettext("This is the Description of the Certificate") %>"><%= description %></textarea>
```
Same comment as for line 18 -- I'm not seeing this placeholder text in the Description field.

catong commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+            </div>
+        </fieldset>
+         <header>
+           <label class="title-2"><%= gettext("Course Signatories") %></label>
```
Don't know whether the heading is coming from here or from cms/templates/js/certificate-details.underscore line 24, but the "Course Signatories" heading in the form is larger than both the "Certificates" and "Certificate Name" headings -- seems too large/not the appropriate level of heading?

cptvitamin replied:
```<label>``` is inappropriate here. Only for form inputs.

catong commented on cms/templates/js/signatory-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+         <fieldset class="collection-fields signatory-fields">
+                <div class="input-wrap field text add-signatory-name">
+                    <label><%= gettext("Name ") %></label>
+                    <input class="collection-name-input input-text signatory-name-input" name="signatory-name" type="text" placeholder="<%= gettext("This is the Name of the Signatory") %>" value="<%= name %>">
```
This placeholder text is not showing up?

catong commented on cms/templates/js/signatory-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+                </div>
+                <div class="input-wrap field text add-signatory-title">
+                    <label><%= gettext("Title ") %></label>
+                    <input class="collection-name-input input-text signatory-title-input" name="signatory-title" type="text" placeholder="<%= gettext("This is the Title of the Signatory") %>" value="<%= title %>">
```
This placeholder text is not showing up?

catong commented on cms/templates/js/signatory-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+                <div class="input-wrap field text add-signatory-name">
+                    <label><%= gettext("Name ") %></label>
+                    <input class="collection-name-input input-text signatory-name-input" name="signatory-name" type="text" placeholder="<%= gettext("This is the Name of the Signatory") %>" value="<%= name %>">
+                    <span class="tip tip-stacked"><%= gettext("Max 30 characters") %></span>
```
If length is not a concern, can we change to "Maximum 30 characters"?

cptvitamin replied:
associate the tip using aria-describedby

catong commented on cms/templates/js/signatory-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+                <div class="input-wrap field text add-signatory-title">
+                    <label><%= gettext("Title ") %></label>
+                    <input class="collection-name-input input-text signatory-title-input" name="signatory-title" type="text" placeholder="<%= gettext("This is the Title of the Signatory") %>" value="<%= title %>">
+                    <span class="tip tip-stacked"><%= gettext("Max 30 characters") %></span>
```
Change to "Maximum 30 characters" if length is not an issue.

catong replied:
I love having this guidance in the UI!
I don't think it's enforced though, is this true? I am able to specify names and titles > 30 chars and save the certificate without causing an error.

cptvitamin replied:
see tip feedback above.

catong commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+         </header>
+        <div class="signatory-edit-list"> </div>
+        <span>
+           <a class="button action-add-signatory" href="javascript:void(0);"><%= gettext("Add Signatory") %></a>
```
@talbs should this look like adding a new group in Group Configurations?

cptvitamin replied:
let's make this a ```<button>``` not a link.  use aria-describedby to link it to the tip below.

catong commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+        <div class="signatory-edit-list"> </div>
+        <span>
+           <a class="button action-add-signatory" href="javascript:void(0);"><%= gettext("Add Signatory") %></a>
+           <span class="tip tip-stacked"><%= gettext("(max 4 signatories are allowed)") %></span>
```
"Up to 4 signatories are allowed for a certificate"

catong commented on cms/static/js/certificates/models/certificate.js:
```diff
@@ -0,0 +1,71 @@
+            // Ensure the provided attributes set meets our expectations for format, type, etc.
+            if (!_.str.trim(attrs.name)) {
+                return {
+                    message: gettext('Certificate name is required.'),
```
I'm not getting this error when I save a certificate without specifying a name.

mattdrayer replied:
@asadiqbal08, can you take a look at the validation framework?

asadiqbal08 replied:
@mattdrayer,  it is strange but i am getting the red colored error while saving a certificate without name.

mattdrayer replied:
Maybe it is a browser-specific issue?

On Tue, May 12, 2015 at 8:18 AM, Asad Iqbal <notifications@github.com>
wrote:

> In cms/static/js/certificates/models/certificate.js
> <https://github.com/edx/edx-platform/pull/7882#discussion_r30127543>:
>
> > +
> > +        setOriginalAttributes: function() {
> > +            // Remember the current state of this model (enables edit->cancel use cases)
> > +            this._originalAttributes = this.parse(this.toJSON());
> > +
> > +            // If no url is defined for the signatories child collection we'll need to create that here as well
> > +            if(!this.isNew() && !this.get('signatories').url) {
> > +                this.get('signatories').url = this.collection.url + '/' + this.get('id') + '/signatories';
> > +            }
> > +        },
> > +
> > +        validate: function(attrs) {
> > +            // Ensure the provided attributes set meets our expectations for format, type, etc.
> > +            if (!_.str.trim(attrs.name)) {
> > +                return {
> > +                    message: gettext('Certificate name is required.'),
>
> @mattdrayer <https://github.com/mattdrayer>, it is strange but i am
> getting the red colored error while saving a certificate without name.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/7882/files#r30127543>.
>



-- 
Matthew Drayer
Solutions Architect
edX | MIT
mattdrayer@edx.org

mattdrayer replied:
I'm okay with removing this code, yes.

On Tue, May 12, 2015 at 9:05 AM, Matthew Drayer <mattdrayer@edx.org> wrote:

> Maybe it is a browser-specific issue?
>
> On Tue, May 12, 2015 at 8:18 AM, Asad Iqbal <notifications@github.com>
> wrote:
>
>> In cms/static/js/certificates/models/certificate.js
>> <https://github.com/edx/edx-platform/pull/7882#discussion_r30127543>:
>>
>> > +
>> > +        setOriginalAttributes: function() {
>> > +            // Remember the current state of this model (enables edit->cancel use cases)
>> > +            this._originalAttributes = this.parse(this.toJSON());
>> > +
>> > +            // If no url is defined for the signatories child collection we'll need to create that here as well
>> > +            if(!this.isNew() && !this.get('signatories').url) {
>> > +                this.get('signatories').url = this.collection.url + '/' + this.get('id') + '/signatories';
>> > +            }
>> > +        },
>> > +
>> > +        validate: function(attrs) {
>> > +            // Ensure the provided attributes set meets our expectations for format, type, etc.
>> > +            if (!_.str.trim(attrs.name)) {
>> > +                return {
>> > +                    message: gettext('Certificate name is required.'),
>>
>> @mattdrayer <https://github.com/mattdrayer>, it is strange but i am
>> getting the red colored error while saving a certificate without name.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/edx/edx-platform/pull/7882/files#r30127543>.
>>
>
>
>
> --
> Matthew Drayer
> Solutions Architect
> edX | MIT
> mattdrayer@edx.org
>



-- 
Matthew Drayer
Solutions Architect
edX | MIT
mattdrayer@edx.org

catong commented on cms/static/js/certificates/views/signatory_editor.js:
```diff
@@ -0,0 +1,131 @@
+            var model = this.model;
+            var self = this;
+            var confirm = new PromptView.Warning({
+                title: gettext('Are you sure you want to delete this signatory with title "'+model.get('title') +'"?'),
```
"Are you sure you want to delete "'+model.get('title') +'" as a signatory?"

catong commented on cms/static/js/certificates/views/certificates_page.js:
```diff
@@ -0,0 +1,54 @@
+        onBeforeUnload: function () {
+            // Check to see if there are any pending changes prior to cancelling and notify the user
+            if (this.certificate.isDirty()) {
+                return gettext('You have unsaved changes. Do you really want to leave this page?');
```
I'm not sure what I need to do to get this error. Tried several things and am not able to produce it :-)

mattdrayer replied:
@asadiqbal08 -- can you take a look at this one and either fix or remove the unused code?

asadiqbal08 replied:
@mattdrayer i think we should remove this code at this time as we are following the relational model concept.

martynjames commented on cms/djangoapps/contentstore/views/certificates.py:
```diff
@@ -0,0 +1,380 @@
+        return certificate_data
+
+    @staticmethod
+    def to_json(certificate):
```
Nit / Optional: `to_json` and `from_json` both seem a little oddly named because strictly speaking they are `to_dict` and `from_dict`

martynjames commented on cms/djangoapps/contentstore/views/certificates.py:
```diff
@@ -0,0 +1,380 @@
+        return certificate
+
+    @staticmethod
+    def deserialize_certificate(course, json_string):
```
Optional: make `deserialize_certificate` a `@classmethod` instead then can call `cls.from_json` - that way if one (anyone) inherits from `CertificateManager` and overrides from_json, the call to `deserialize_certificate` will do the right thing.

martynjames commented on cms/djangoapps/contentstore/views/certificates.py:
```diff
@@ -0,0 +1,380 @@
+        Retrieve the certificates list from the provided course
+        """
+        # The top-level course field is 'certificates'
+        course_certificates_field = getattr(course, 'certificates')
```
if a plain `getattr` fails then the exception is the same as if you'd just asked for the attribute value normally:
```
class Y(object):
   def __init__(self, a, b):
     self.a = a
     self.b = b
y = Y(12,4)

>>> y.c
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Y' object has no attribute 'c'
>>> getattr(y, 'c')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Y' object has no attribute 'c'
>>> 
```

As a result I think that this whole thing could be written as:
`return course.certificates.get('certificates', [])`

martynjames commented on cms/djangoapps/contentstore/views/certificates.py:
```diff
@@ -0,0 +1,380 @@
+                course.certificates['certificates'].pop(index)
+                store.update_item(course, request.user.id)
+                break
+        return JsonResponse(status=204)
```
None of the other methods (except `remove_signatory` below) return a `JsonResponse` - Seems like the calling view should be the thing to return the `JsonResponse`

martynjames commented on cms/djangoapps/contentstore/views/certificates.py:
```diff
@@ -0,0 +1,380 @@
+        self.certificate_data = certificate_data
+        self.id = certificate_data['id']  # pylint: disable=invalid-name
+
+    def get_certificate_data(self):
```
Could make this a @property, then I think one can remove the `# pylint: disable=no-member` flag when calling above... although I'm not even sure that it's needed anyway (I don't appear to get the error using a local pylint plugin)

martynjames commented on cms/djangoapps/contentstore/views/certificates.py:
```diff
@@ -0,0 +1,380 @@
+            certificate_url = reverse_course_url('certificates.certificates_list_handler', course_key)
+            course_outline_url = reverse_course_url('course_handler', course_key)
+            certificates = None
+            print 'VIEW SETTING: {}'.format(settings.FEATURES.get('CERTIFICATES_HTML_VIEW'))
```
Not sure that this (and below) `print` statement is particularly useful - maybe `log.debug` instead?

martynjames commented on cms/djangoapps/contentstore/views/certificates.py:
```diff
@@ -0,0 +1,380 @@
+                certificates_list.append(CertificateManager.to_json(new_certificate))
+
+            store.update_item(course, request.user.id)
+            return JsonResponse(CertificateManager.to_json(new_certificate), status=201)
```
We have already perfrmed the `CertificateManager.to_json`above - might be worth assigning it to a variable and using it in 2 places.

martynjames commented on cms/djangoapps/contentstore/views/certificates.py:
```diff
@@ -0,0 +1,380 @@
+    """
+    course_key = CourseKey.from_string(course_key_string)
+    store = modulestore()
+    with store.bulk_operations(course_key):
```
Not sure that you need `bulk_operations` here, seems to be only the course-load that uses it. Shouldn't matter, but when there are only fetches, I'm not certain that it is necessary...

mattdrayer replied:
removed the use of bulk_operations here, for now

martynjames commented on cms/static/coffee/spec/main.coffee:
```diff
@@ -28,6 +28,7 @@ requirejs.config({
         "backbone": "xmodule_js/common_static/js/vendor/backbone-min",
         "backbone.associations": "xmodule_js/common_static/js/vendor/backbone-associations-min",
         "backbone.paginator": "xmodule_js/common_static/js/vendor/backbone.paginator.min",
+        "backbone-relational": "xmodule_js/common_static/js/vendor/backbone-relational",
```
We should probably be using minified version - `backbone-relational.min`

martynjames commented on cms/static/js/certificates/collections/certificates.js:
```diff
@@ -0,0 +1,41 @@
+        parse: function (certificatesJson) {
+            // Transforms the provided JSON into a Certificates collection
+            var modelArray;
+            if(this.IsJsonString(certificatesJson)) {
```
This approach means that you're calling `JSON.parse` twice on any string - why not test if the value is a string - if a string, then `JSON.parse`, otherwise assume it's already a json object.

Alternatively, could have the translation / non-translation happen inside some wrapped function - e.g.
```
certificate_array: function(certificate_info) {
    var return_array;
    try {
        return_array = JSON.parse(certificate_info);
    } catch(e) {
        return_array = certificate_info;
    }
    return return_array;
}
```

martynjames commented on cms/static/js/certificates/models/certificate.js:
```diff
@@ -0,0 +1,71 @@
+],
+function(Backbone, Signatory, SignatoryCollection, _, str, gettext) {
+    'use strict';
+    _.str = str;
```
Not sure I understand why this is important - doesn't this just make `_.str` equivalent to `_.string`?

mattdrayer replied:
I took a look at this and it seems Backbone is looking for the _.str varialble -- @asadiqbal08 , can you provide further explanation?

asadiqbal08 replied:
_.str is underscore sting object.

martynjames replied:
Yes I see that, but it looks like it is imported as underscore.string anyway - is that a separate js library? then it makes sense. It's not a criticism, more a desire to know.

asadiqbal08 replied:
underscore.string is reflected at this path  js/vendor/underscore.string.min , 
you can see the mapping in require-config.js as well.

martynjames commented on cms/static/js/certificates/spec/views/certificate_details_spec.js:
```diff
@@ -0,0 +1,114 @@
+) {
+    'use strict';
+
+    var SELECTORS = {
```
Love this!

martynjames commented on cms/static/js/certificates/views/certificate_details.js:
```diff
@@ -0,0 +1,71 @@
+
+        editCertificate: function(event) {
+            // Flip the model into 'editing' mode
+            if (event && event.preventDefault) { event.preventDefault(); }
```
If it's really a javascript `event` then `preventDefault` should be present - should do no harm, and allows one to call these functions with a "fake" event if desired (maybe for testing...)

martynjames commented on cms/static/js/certificates/views/certificate_details.js:
```diff
@@ -0,0 +1,71 @@
+                'collection',
+                'certificates',
+                'certificate-details',
+                'certificate-details-' + index
```
Not sure of the need for this indexy class name... I may be mistaken, but there does not appear to be any styles / js code using it. My first reaction was that there are style-rules that could let you change the style according to position-index in a list, instead of appending an index to the className.

cptvitamin commented on cms/templates/certificates.html:
```diff
@@ -0,0 +1,86 @@
+            </div>
+          % else:
+            <div class="ui-loading">
+                <p><span class="spin"><i class="icon fa fa-refresh"></i></span> <span class="copy">${_("Loading")}</span></p>
```
@mattdrayer I'm recommending that, until we get rid of icon fonts altogether, we put an ```aria-hidden="true"``` attribute on an ```<i>``` used to render an icon: https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Accessibility#using-icon-fonts

cptvitamin commented on cms/templates/certificates.html:
```diff
@@ -0,0 +1,86 @@
+    <section class="content">
+      <article class="content-primary" role="main">
+        <div class="wrapper-certificates certificates-list">
+          <h3 class="title">${_("Certificates")}</h3>
```
Looks like this is a child of the ```<h1>``` used in the header above.  Either we change that to an ```<h2>``` or change the ```<h3>``` 's below to ```<h2>```'s.  I'm having trouble pulling up the sandbox right now, so I don't know where this appears in the context of the page.  I'm guessing this page already has an ```<h1>``` so I recommend changing the page header to use an ```<h2>```

cptvitamin commented on cms/templates/certificates.html:
```diff
@@ -0,0 +1,86 @@
+            advanced_settings_url = utils.reverse_course_url('advanced_settings_handler', context_course.id)
+          %>
+        <h3 class="title-3">${_("Other Course Settings")}</h3>
+          <nav class="nav-related" aria-label="${_('Other Course Settings')}">
```
Bonus points for including an aria-label :+1:

cptvitamin commented on cms/templates/js/certificate-details.underscore:
```diff
@@ -0,0 +1,40 @@
+<div class="collection-details wrapper-certificate">
+    <header class="collection-header">
+        <h3 class="title">
+            <a href="#" class="toggle detail-toggle <% if (showDetails){ print('hide'); } else { print('show'); } %>-details">
```
```href="#"``` is actually not valid.  It's usually a sure sign that this should be a ```<button>``` and not a hyperlink.  (Buttons do something, Links go somewhere).  I would make this a button, which means it probably shouldn't be wrapped in an h3.

mattdrayer replied:
@asadiqbal08 -- if you can work on switching these anchors/links over to buttons, that would be great.

marcotuts replied:
Lined 4 and 5 should be removed - no need to have collapsed state for this item.

cptvitamin commented on cms/templates/js/certificate-details.underscore:
```diff
@@ -0,0 +1,40 @@
+    <header class="collection-header">
+        <h3 class="title">
+            <a href="#" class="toggle detail-toggle <% if (showDetails){ print('hide'); } else { print('show'); } %>-details">
+                <i class="ui-toggle-expansion icon fa fa-caret-<% if (showDetails){ print('down'); } else { print('right'); } %>"></i>
```
add an ```aria-hidden="true"```

cptvitamin commented on cms/templates/js/certificate-details.underscore:
```diff
@@ -0,0 +1,40 @@
+        </li>
+        <li class="collection-description certificate-description">
+         <header>
+           <label class="title-2"><%= gettext("Course Signatories") %></label>
```
```<label>``` elements are only for form inputs.  Should probably use some other element here.

cptvitamin commented on cms/templates/js/certificate-details.underscore:
```diff
@@ -0,0 +1,40 @@
+
+    <ul class="actions certificate-actions">
+        <li class="action action-edit">
+            <button class="edit"><i class="icon fa fa-pencil"></i> <%= gettext("Edit") %></button>
```
icon font, hide it from sr users.

cptvitamin commented on cms/templates/js/certificate-details.underscore:
```diff
@@ -0,0 +1,40 @@
+            <button class="edit"><i class="icon fa fa-pencil"></i> <%= gettext("Edit") %></button>
+        </li>
+        <li class="action action-delete wrapper-delete-button" data-tooltip="<%= gettext('Delete') %>">
+            <button class="delete action-icon"><i class="icon fa fa-trash-o"></i><span><%= gettext("Delete") %></span></button>
```
hide it.

cptvitamin commented on cms/templates/js/certificate-details.underscore:
```diff
@@ -0,0 +1,40 @@
+      <% } %>
+   </ol>
+
+    <ul class="actions certificate-actions">
```
Is there a reason why this is an unordered list?  I'm not a fan of using ```<ul>``` or ```<ol>``` for anything but lists of items.  They add a lot of verbosity for screen reader users.  While they are useful for lists of items, they actually become cumbersome in other contexts.

mattdrayer replied:
I see no reason why we need to use lists here -- @asadiqbal08, can you address?

cptvitamin commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+<form class="collection-edit-form certificate-edit-form">
+    <% if (error && error.message) { %>
+        <div class="certificate-edit-error message message-status message-status error is-shown" name="certificate-edit-error">
```
A screen reader user will not be aware of any message that appears here since their screen reader will only read what is currently in focus.   ```aria-live``` attribute to the rescue, it tells the screen reader to read text that gets injected into the DOM inside the element with the attribute.  Problem with ```aria-live``` is that it cannot be used on the node that is being injected, it has to be used on a node that already exists in the DOM.  Can you create a parent ```<div aria-live="polite">``` here, outside the conditional?  Anything that gets added to that div, like this error message, will be read to SR user as soon as the SR stops talking.

cptvitamin commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+    <% } %>
+    <div class="wrapper-form">
+        <fieldset class="collection-fields certificate-fields">
+            <legend class="sr"><%= gettext("Certificate Information") %></legend>
```
nice

cptvitamin commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+            <div class="input-wrap field text add-certificate-description">
+                <label for="certificate-description-<%= uniqueId %>"><%= gettext("Description") %></label>
+                <textarea id="certificate-description-<%= uniqueId %>" class="certificate-description-input text input-text" name="certificate-description" placeholder="<%= gettext("This is the Description of the Certificate") %>"><%= description %></textarea>
+                <span class="tip tip-stacked"><%= gettext("Optional long description") %></span>
```
see tip recommendation above

cptvitamin commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+        <button class="action action-secondary action-cancel"><%= gettext("Cancel") %></button>
+        <% if (!isNew) { %>
+            <span class="wrapper-delete-button">
+                <a class="button action-delete delete" href="#"><%= gettext("Delete") %></a>
```
This should be a ```<button>```

mattdrayer replied:
@asadiqbal08 -- another anchor-to-button job

talbs replied:
@mattdrayer and @asadiqbal08, I'll take care of this quick change.

Thanks for the catch, @cptvitamin!

mattdrayer replied:
@talbs -- any chance you can work your UI magic on this anchor tag as well?  Needs to be converted to a button.

cptvitamin commented on cms/templates/js/signatory-details.underscore:
```diff
@@ -0,0 +1,18 @@
+     <div class="signatory-panel-default">
+       <div class="actions certificate-actions signatory-panel-edit">
+        <span class="action action-edit-signatory">
+            <a href="javascript:void(0);" class="edit-signatory"><i class="icon fa fa-pencil"></i> <%= gettext("Edit") %></a>
```
This should be a button and the icon should be aria-hidden

cptvitamin commented on cms/templates/js/signatory-details.underscore:
```diff
@@ -0,0 +1,18 @@
+       <div class="signatory-panel-header">Signatory <%= signatory_number %></div>
+       <div class="signatory-panel-body">
+                <div>
+                    <label><%= gettext("Name: ") %></label>
```
labels must only be used for form inputs.

cptvitamin commented on cms/templates/js/signatory-details.underscore:
```diff
@@ -0,0 +1,18 @@
+            <a href="javascript:void(0);" class="edit-signatory"><i class="icon fa fa-pencil"></i> <%= gettext("Edit") %></a>
+        </span>
+        </div>
+       <div class="signatory-panel-header">Signatory <%= signatory_number %></div>
```
Should this be a heading?

cptvitamin commented on cms/templates/js/signatory-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+     <div class="signatory-panel-default">
+
+       <% if (!is_editing_all_collections) { %>
+       <a class="signatory-panel-close" href="javascript:void(0);" data-tooltip="Close">
```
all of these should be buttons not links.

mattdrayer replied:
@asadiqbal08, FYI

talbs replied:
@mattdrayer and @asadiqbal08, I'll take care of this quick change.

Thanks for the catches, @cptvitamin!

mattdrayer replied:
@talbs -- some more link-to-button spots

cptvitamin commented on cms/templates/js/signatory-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+       <% } %>
+       <div class="signatory-panel-header">Signatory <%= signatory_number %></div>
+       <div class="signatory-panel-body">
+         <fieldset class="collection-fields signatory-fields">
```
fieldsets must have a ```<legend>``` child that describe the collection.

mattdrayer replied:
added!

cptvitamin commented on cms/templates/js/signatory-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+       <div class="signatory-panel-body">
+         <fieldset class="collection-fields signatory-fields">
+                <div class="input-wrap field text add-signatory-name">
+                    <label><%= gettext("Name ") %></label>
```
If the associated ```<input>``` is not a child of the ```<label>``` element, then the label element needs a ```for``` attribute that references the ID of the input.  It's usually easier just to make the input a child.

cptvitamin commented on cms/templates/js/signatory-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+                    <span class="tip tip-stacked"><%= gettext("Max 30 characters") %></span>
+                </div>
+                <div class="input-wrap field text add-signatory-title">
+                    <label><%= gettext("Title ") %></label>
```
see label feedback above.

martynjames commented:
Some issues found while testing:
* We are allowed to enter more than 30 characters in fields that say they have a maximum of 30 characters
* Upon revisiting Certificates page, the views for existing certificates are in edit mode - seems that we should be opening them in view mode.
* Having a single-quote character ( ' ) in a certificate body is problematic, and causes javascript to fail on the page when reopened. e.g. Try a description of "It's easy as 123". Probably need to escape the json for the `CertificatesPageFactory` initialization

mattdrayer commented:
Okay, i've addressed a ton of the feedback provided during the first round review of this PR.  

Regarding @martynjames' specific comments at the end of his initial reivew, I've added basic 'maxlength' field constraints to those fields limiting entries to 30 characters -- @asadiqbal08 can provide additional validation logic, if necessary.  Also, @asadiqbal08, can you take a look at the misbehaviors noted by @martynjames -- page rendering in 'edit' mode after refresh, and single-apostrophe causing issues with Javascript.  Thanks!

ziafazal commented:
Jenkins, test this please.

mattdrayer commented:
FYI, sandbox is currently being updated -- should be available in about 30 mins

marcotuts commented:
@mattdrayer All of the -extends sass files are back it seems. I don't know if my commit got lost in a rebase + gif push -f perhaps? I can make the change again, but I wanted to check to see if it wasn't something else.

cahrens commented:
I am experiencing odd behavior with Tab navigation in the Studio form to create a new certificate (Chrome). Repro steps:

1) Delete placeholder text in Signatory name field
2) Click in a different field
3) Click back in name field and type something
4) Press Tab to move to the next field-- focus does not move there and seems to go to the top of the window.

cahrens commented:
Wow, this save icon is *really* retro. I wonder what percent of our students know what a floppy disk is?

![image](https://cloud.githubusercontent.com/assets/484484/7659485/41dd268c-fb10-11e4-8451-fb2f619100d2.png)

cahrens commented:
The "Learn More" button on the RHS does not seem to go to a specific doc location. Should it?

cahrens commented:
The "delete signatory" confirmation shows title instead of name. 

![image](https://cloud.githubusercontent.com/assets/484484/7659591/119791be-fb11-11e4-9ca0-9db6b5c84383.png)

cahrens commented:
Group Configurations is missing from this list of other Settings page (RHS)

![image](https://cloud.githubusercontent.com/assets/484484/7659674/acfbff3c-fb11-11e4-931d-6ff7c6b0f013.png)

cahrens commented:
Don't forget to update on-page help documentation.

![image](https://cloud.githubusercontent.com/assets/484484/7659823/d31df85e-fb12-11e4-83b7-7af5d725da4d.png)

cahrens commented:
There are some Jasmine helper methods introduced in this PR that are showing no code coverage. Does this mean that a spec is not getting run?

https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/6359/Diff_Coverage_Report/

cahrens commented on cms/djangoapps/contentstore/views/certificates.py:
```diff
@@ -0,0 +1,374 @@
+    """
+    course_key = CourseKey.from_string(course_key_string)
+    store = modulestore()
+    with store.bulk_operations(course_key):
```
I question if the bulk_operations usages are necessary. @cpennington can you offer advise on whey they should be used?

cahrens commented on cms/djangoapps/contentstore/views/exception.py:
```diff
@@ -0,0 +1,18 @@
+"""
+A common module for managing exceptions. Helps to avoid circular references
```
It appears to me that the only reference to CertificateValidationError is in views/certificates.py (and there are no usages of CertificateException beyond CertificateValidationError). Can we just put the exception class directly in that file?

mattdrayer replied:
@ziafazal, ref: SOL-907

cahrens commented on cms/static/coffee/spec/main.coffee:
```diff
@@ -268,6 +272,11 @@ define([
     "js/spec/xblock/cms.runtime.v1_spec",
 
+    # Certificates application test suite mappings
+    "js/certificates/spec/models/certificate_spec",
```
There are more new specs that need to be added to this list.

mattdrayer replied:
@ziafazal, ref: SOL-907

cahrens commented on lms/templates/certificates/_certificate-signatories.html:
```diff
@@ -0,0 +1,14 @@
+<div>
+   <span style="margin-bottom: 15px; display: block; font-weight: 600">
+       ${'Certificate Signatories:'}
```
Does this need to be internationalized?

cahrens commented on common/test/acceptance/tests/studio/test_studio_settings.py:
```diff
@@ -397,3 +398,217 @@ def test_only_expected_fields_are_displayed(self):
         self.assertEquals(set(displayed_fields), set(expected_fields))
+
+
+class CertificatesTest(StudioCourseTest):
```
Nit: it might be nice to put this test in its own file.

cahrens commented on common/test/acceptance/tests/studio/test_studio_settings.py:
```diff
@@ -397,3 +398,217 @@ def test_only_expected_fields_are_displayed(self):
+        And I set new the name, description and signatory and click the button 'Create'
+        Then I see the new certificate is added and has one signatory inside it
+        When I click 'Edit' button of signatory panel
+        And I set the name and click the button 'Save' icon
```
It would be nice to have a test around deleting a signatory.

cahrens commented on common/test/acceptance/tests/studio/test_studio_settings.py:
```diff
@@ -397,3 +398,217 @@ def test_only_expected_fields_are_displayed(self):
+        certificate = self.certificates_page.certificates[0]
+        certificate.name = ""
+        certificate.save()
+        self.assertEqual(certificate.mode, 'edit')
```
If validation of signatory fields is not tested in a unit test, it would be good to add a bok choy test. If it is tested via unit tests (Jasmine), no need to duplicate.

cahrens commented on cms/static/js/certificates/collections/certificates.js:
```diff
@@ -0,0 +1,37 @@
+            try {
+                return_array = JSON.parse(certificate_info);
+            } catch (e) {
+                return_array = certificate_info;
```
Can you add a comment about when you would expect to get into this case?

mattdrayer replied:
@asadiqbal08 -- can you add some context around the catch condition in this code block?

cahrens commented on cms/static/js/certificates/factories/certificates_page_factory.js:
```diff
@@ -0,0 +1,31 @@
+to create a page object.  This was added for the RequireJS Optimizer, which needs to have a single root to determine
+statically all of the dependencies needed by a page.  The RequireJS Optimizer combines these dependencies into a single
+optimized JS file. Mako templates typically contain a block that constructs the page object using this page factory.
+Unit tests for the page factory verify that it behaves as desired. Some of these factories are more complex than others.
```
I do not see (on sandbox) that the new JS files are getting optimized into a single file. Add your factory file to the list in cms/static/build.js.

mattdrayer replied:
@ziafazal, ref: SOL-907

cahrens commented on cms/static/js/certificates/models/certificate.js:
```diff
@@ -0,0 +1,81 @@
+    var Certificate = Backbone.RelationalModel.extend({
+        idAttribute: "id",
+        defaults: {
+            name: 'Name of the certificate',
```
If these defaults are not going away (I believe they are!), then they need to be internationalized.

mattdrayer replied:
@asadiqbal08 -- after SOL-386 is merged we can fix the placeholder text issue

cahrens commented on cms/static/js/certificates/models/certificate.js:
```diff
@@ -0,0 +1,81 @@
+                return {
+                    message: gettext('Signatory field(s) has invalid data.'),
+                    attributes: {signatories: attrs.signatories.models}
+                }
```
Missing semi-colon (note that syntax highlighting is off after this).

mattdrayer replied:
@ziafazal, ref: SOL-907

cahrens commented on cms/static/js/certificates/factories/certificates_page_factory.js:
```diff
@@ -0,0 +1,31 @@
+    'js/certificates/collections/certificates', 'js/certificates/models/certificate', 'js/certificates/views/certificates_page'
+], function(CertificatesCollection, Certificate, CertificatesPage) {
+    'use strict';
+    return function (certificatesJson, certificateUrl, courseOutlineUrl) {
```
courseOutlineUrl not used

cahrens commented on cms/static/js/certificates/factories/certificates_page_factory.js:
```diff
@@ -0,0 +1,31 @@
+        });
+        // Execute the page object's rendering workflow
+        var certificatesPage = new CertificatesPage({
+            el: $('#content'),
```
Declare dependency on JQuery.

mattdrayer replied:
I've updated all of the RequireJS dependency declarations in this new app, so we should be all set here, now

cahrens commented on cms/static/js/certificates/models/signatory.js:
```diff
@@ -0,0 +1,69 @@
+define([
+    'backbone', 'underscore', 'underscore.string', 'gettext', 'backbone-relational'
+],
+function(Backbone, _, str) {
```
Add gettext variable.

cahrens commented on cms/static/js/certificates/models/signatory.js:
```diff
@@ -0,0 +1,69 @@
+                    }, errors);
+                }
+                else if ((lines.length > 1 && (lines[0].length > 40 || lines[1].length > 40))
+                    || (lines.length == 1 && title.length > 40)) {
```
Bad line break and ==. I'd suggest you lint your JS.

mattdrayer replied:
@ziafazal, ref: SOL-907 -- we'll cover this as part of the linting run we're conducting with the new JSHint utility available in devstack

cahrens commented on cms/static/js/certificates/views/certificate_details.js:
```diff
@@ -0,0 +1,70 @@
+define([
+    'js/views/baseview', 'underscore', 'js/certificates/models/signatory', 'js/certificates/views/signatory_details', 'gettext', 'underscore.string'
+],
+function(BaseView, _, Signatory, SignatoryDetails, gettext, str) {
```
gettext, str not used

cahrens commented on cms/static/js/certificates/views/certificate_details.js:
```diff
@@ -0,0 +1,70 @@
+
+        className: function () {
+            // Determine the CSS class names for this model instance
+            var index = this.model.collection.indexOf(this.model);
```
This variable is not used.

mattdrayer replied:
@asadiqbal08 -- FYI on this unused variable (scoped to function) -- did you intend to utilize it for something?  It could be a holdover from the group configurations bootstrap we did early-on.

cahrens commented on cms/static/js/certificates/views/certificate_editor.js:
```diff
@@ -0,0 +1,128 @@
+define(['js/views/list_item_editor', 'js/certificates/models/signatory', 'js/certificates/views/signatory_editor', 'underscore', 'jquery', 'gettext'],
+function(ListItemEditorView, Signatory, SignatoryEditor, _, $, gettext) {
+    'use strict';
+    var MIN_SIGNATORIES_LIMIT = 1;
```
Unused variable.

cahrens commented on cms/static/js/certificates/views/certificate_editor.js:
```diff
@@ -0,0 +1,128 @@
+        initialize: function() {
+            // Set up the initial state of the attributes set for this model instance
+            _.bindAll(this, "onSignatoryRemoved", "clearErrorMessage");
+            this.eventAgg = _.extend({}, Backbone.Events);
```
Backbone not defined.

mattdrayer replied:
@ziafazal, ref: SOL-907

cahrens commented on cms/static/js/certificates/views/certificate_item.js:
```diff
@@ -0,0 +1,40 @@
+    'use strict';
+    var CertificateItemView = ListItemView.extend({
+        events: {
+            'click .delete': 'deleteItem'
```
I don't see deleteItem defined.

cahrens commented on cms/static/js/certificates/views/signatory_editor.js:
```diff
@@ -0,0 +1,144 @@
+            var model = this.model;
+            var self = this;
+            var confirm = new PromptView.Warning({
+                title: gettext('Are you sure you want to delete "'+model.get('title') +'" as a signatory?'),
```
You can't do concatenate strings like this for translation. You need to use a placeholder and the interpolate function.

http://edx-developer-guide.readthedocs.org/en/latest/internationalization/i18n.html

cahrens commented on cms/static/js/certificates/views/certificates_page.js:
```diff
@@ -0,0 +1,41 @@
+
+        expandCertificate: function (id) {
+            // Locate the certificate having a corresponding identifier
+            var certificate = this.certificates.findWhere({
```
This method doesn't seem to be doing anything.

cahrens commented on cms/static/js/spec_helpers/view_helpers.js:
```diff
@@ -94,6 +94,41 @@ define(["jquery", "js/views/feedback_notification", "js/views/feedback_prompt",
+            ViewHelpers.verifyNotificationShowing(notificationSpy, /Saving/);
+        };
+
+
```
Nit: Extra newlines.

cahrens commented on cms/templates/js/signatory-details.underscore:
```diff
@@ -0,0 +1,22 @@
+  <div class="signatory-panel-header">Signatory <%= signatory_number %></div>
+  <div class="signatory-panel-body">
+    <div>
+      <span class="signatory-name-label"><%= gettext("Name: ") %></span>
```
I would suggest a translator note that it is important to keep the space. It would be better though if you could do this as a single string with a placeholder to make translation more robust. Not sure if that is possible with the classes.

cahrens commented:
A couple of parting thoughts-- make sure the UI is tested on all supported browsers (thankfully IE9 is out now!), and if you haven't done performance testing of the new Studio page, ask to see if it is necessary. I know all new LMS pages need performance testing, but I am not sure about our position for Studio pages.

marcotuts commented on cms/templates/certificates.html:
```diff
@@ -0,0 +1,86 @@
+    <section class="content">
+      <article class="content-primary" role="main">
+        <div class="wrapper-certificates certificates-list">
+          <h2 class="title">${_("Certificates")}</h2>
```
Remove this title / h2, unless other pages in studio also keep this pattern, in which case add the "sr" class to hide it visually.

marcotuts commented on cms/static/sass/views/_certificates.scss:
```diff
@@ -0,0 +1,723 @@
+          margin-right: ($baseline*14);
+          color: $black;
+
+          .toggle {
```
remove styles for toggling from this view (lines 58 -->93 or so) This also should be removed from HTML.

marcotuts commented on cms/templates/js/certificate-details.underscore:
```diff
@@ -0,0 +1,40 @@
+
+   <ol class="collection-info certificate-info certificate-info-<% if(showDetails){ print('block'); } else { print('inline'); } %>">
+     <% if (!_.isUndefined(id)) { %>
+         <li class="certificate-id">
```
No need to render certificate ID in this view, this list item can be removed.

marcotuts commented on cms/templates/js/certificate-details.underscore:
```diff
@@ -0,0 +1,40 @@
+            <span class="certificate-value"><%= id %></span>
+        </li>
+     <% } %>
+     <% if (showDetails) { %>
```
No need to have certificate descriptions, this list item can be removed. (and potentially the associated field, unless we want to default it to something else or keep it empty.)

mattdrayer replied:
I've removed the certificate description field from the UI -- I'd like to keep it as part of the model for now, however.

marcotuts commented on cms/templates/js/certificate-details.underscore:
```diff
@@ -0,0 +1,40 @@
+        </li>
+        <li class="collection-description certificate-description">
+         <header>
+           <span class="title-2"><%= gettext("Course Signatories") %></span>
```
use same class as the edit view for certificates (shown below), remove header wrapper from this area.
![image](https://cloud.githubusercontent.com/assets/2023680/7681458/70f64482-fd3e-11e4-8c77-18cd8056c8bc.png)

marcotuts commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,47 @@
+            <div class="input-wrap field text required add-collection-name add-certificate-name <% if(error && error.attributes && error.attributes.name) { print('error'); } %>">
+                <label for="certificate-name-<%= uniqueId %>"><%= gettext("Certificate Name") %></label>
+                <% if (!_.isUndefined(id)) { %>
+                    <span class="certificate-id">
```
Certificate ID can be removed from the editing view as well.

marcotuts commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,47 @@
+                        <span class="certificate-value"><%= id %></span>
+                    </span>
+                <% } %>
+                <input id="certificate-name-<%= uniqueId %>" class="collection-name-input input-text" name="certificate-name" type="text" placeholder="<%= gettext("Name of the certificate") %>" value="<%= name %>" aria-describedby="certificate-name-<%=uniqueId %>-tip">
```
The name of the certificate should be removed and instead replaced with other general information fields. (If those fields are in a separate PR, it would be great to remove the need to set a name field entirely and hide this from the editing view.)

marcotuts commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,47 @@
+                <input id="certificate-name-<%= uniqueId %>" class="collection-name-input input-text" name="certificate-name" type="text" placeholder="<%= gettext("Name of the certificate") %>" value="<%= name %>" aria-describedby="certificate-name-<%=uniqueId %>-tip">
+                <span id="certificate-name-<%= uniqueId %>-tip" class="tip tip-stacked"><%= gettext("Name of the certificate") %></span>
+            </div>
+            <div class="input-wrap field text add-certificate-description">
```
Certificate description should also be removed from the editing view

marcotuts commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,47 @@
+         </header>
+        <div class="signatory-edit-list"> </div>
+        <span>
+           <button class="action action-add-signatory" type="button"><%= gettext("Add Signatory") %></button>
```
Styles seem to be broken for this button action currently. This should use the same styling as the add / create actions at the bottom of lists, as shown below:

![image](https://cloud.githubusercontent.com/assets/2023680/7681514/dc64e5fc-fd3e-11e4-9c2a-423b40d6a0ea.png)

marcotuts commented on cms/templates/js/certificate-editor.underscore:
```diff
@@ -0,0 +1,47 @@
+        </span>
+    </div>
+    <div class="actions">
+        <button class="action action-primary" type="submit"><% if (isNew) { print(gettext("Create")) } else { print(gettext("Save")) } %></button>
```
If we'd like to keep signatory panel editing, I would suggest using the same save / cancel / delete pattern we use elsewhere in Studio instead of this save icon and delete icon during edit mode.  

Example: 
![image](https://cloud.githubusercontent.com/assets/2023680/7681586/4ae95eea-fd3f-11e4-8f0c-31f3ac4ea771.png)

mattdrayer replied:
@marcotuts -- if you have the time it would be great to get a hand on this particular item so we can focus on finishing up SOL-386

marcotuts commented on cms/static/sass/views/_certificates.scss:
```diff
@@ -0,0 +1,723 @@
+
+      .collection-info {
+        @extend %t-copy-sub1;
+        color: $gray-l1;
```
remove this margin left and color rule (pairs with removal of expand / collapse action for this area.

marcotuts commented on cms/static/sass/views/_certificates.scss:
```diff
@@ -0,0 +1,723 @@
+          }
+
+          .name {
+            overflow: hidden;
```
test out using the %cont-truncated extend instead

marcotuts commented:
I haven't reviewed the styling in full detail, but I thought it made more sense to do that detailed review after you've had a chance to review my comments so far. Let me know if you have any questions!

marcotuts commented:
One other note - We should only allow for the creation of 1 certificate, and disable / hide the "New Certificate" button after one 'certificate' has been added.

marcotuts commented:
Additional changes necessary prior to merge:

- [ ] review / double-check RTL styling 

Additional changes from a FED standpoint that would be nice to include in this PR. If these items wind up blocking merge, we should create a backlog story to address them. For now, still presuming UX support task includes this work! 

- [ ] update save / cancel icons for per-signatory editing (this is related to @cptvitamin's feedback about using buttons instead of links. rather than updating these current icons to be buttons, the hope is to use a similar save / cancel bar pattern for just editing 1 signatory at a time. 
- [ ] clean up _certificates.scss to use as much as it can from _collections.scss, and more clearly denote when it is overriding those styles. This is a bit muddled right now.

marcotuts commented on lms/static/certificates/sass/_components.scss:
```diff
@@ -0,0 +1,564 @@
+        }
+
+        .icon {
+            margin-right: spacing-horizontal(x-small);
```
noob question for @talbs about pattern library styling here. - Is this something we'd need to wrap with our handy RTL mixins or is that handled some other way in our PL work? 

I am opening up a PR with RTL fixes to merge back into this branch, and wanted to check about the PL files before doing any work here. 

Thanks!

talbs replied:
@marcotuts, yeah, we definitely need this to be RTL friendly with our syntax. Thanks for the catch. If you haven't started yet, I can make that RTL PR (since I should have done this to begin with).

Let me know where you are at. Happy to take this on as part of my FED support on this feature.

marcotuts replied:
@talbs - Yeah if you wouldn't mind taking care of that, that would be great! I've merged in RTL support for the studio settings page, but I didn't touch the web view work.

marcotuts commented:
I have a PR that will add RTL support for this view, and concluedes the "must have" list from my perspective for merge. (https://github.com/edx/edx-platform/pull/8258)  Once this PR is merged I'll update my task list above.

marcotuts commented on cms/static/sass/views/_certificates.scss:
```diff
@@ -0,0 +1,726 @@
+    .collection-details {
+
+      .actions {
+        @include transition(opacity .15s .25s ease-in-out);
```
As @clrux  noted in a separate PR, this is out of order. Should be @include transition(opacity .15s ease-in-out .25s);

marcotuts commented:
I've merged in the RTL fixes to the Studio Certificates setting page from a separately reviewed PR. My list of must fixes from a FED standpoint is complete, so I'm a thumbs up on this. :+1: 

The list of improvements I included above we can create a backlog story for, as well as the RTL work @talbs volunteered to help with for the web view. 

Thanks everyone!
